// xdl/build/xdl.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Analytics", {
  enumerable: true,
  get: function () {
    return _Analytics().default;
  }
});
Object.defineProperty(exports, "Api", {
  enumerable: true,
  get: function () {
    return _Api().default;
  }
});
Object.defineProperty(exports, "ApiV2", {
  enumerable: true,
  get: function () {
    return _ApiV().default;
  }
});
Object.defineProperty(exports, "Config", {
  enumerable: true,
  get: function () {
    return _Config().default;
  }
});
Object.defineProperty(exports, "ErrorCode", {
  enumerable: true,
  get: function () {
    return _ErrorCode().ErrorCode;
  }
});
Object.defineProperty(exports, "Logger", {
  enumerable: true,
  get: function () {
    return _Logger().default;
  }
});
Object.defineProperty(exports, "NotificationCode", {
  enumerable: true,
  get: function () {
    return _NotificationCode().default;
  }
});
Object.defineProperty(exports, "UserManager", {
  enumerable: true,
  get: function () {
    return _User().default;
  }
});
Object.defineProperty(exports, "RobotUser", {
  enumerable: true,
  get: function () {
    return _User().RobotUser;
  }
});
Object.defineProperty(exports, "User", {
  enumerable: true,
  get: function () {
    return _User().User;
  }
});
Object.defineProperty(exports, "UserSettings", {
  enumerable: true,
  get: function () {
    return _UserSettings().default;
  }
});
Object.defineProperty(exports, "XDLError", {
  enumerable: true,
  get: function () {
    return _XDLError().default;
  }
});
Object.defineProperty(exports, "PackagerLogsStream", {
  enumerable: true,
  get: function () {
    return _PackagerLogsStream().default;
  }
});
Object.defineProperty(exports, "LogRecord", {
  enumerable: true,
  get: function () {
    return _PackagerLogsStream().LogRecord;
  }
});
Object.defineProperty(exports, "LogUpdater", {
  enumerable: true,
  get: function () {
    return _PackagerLogsStream().LogUpdater;
  }
});
exports.ModuleVersion = exports.ProjectUtils = exports.Doctor = exports.PKCS12Utils = exports.AndroidCredentials = exports.Webpack = exports.Versions = exports.UrlUtils = exports.ThirdParty = exports.StandaloneBuild = exports.Simulator = exports.SimControl = exports.Prompts = exports.ProjectSettings = exports.ProjectAssets = exports.Project = exports.Exp = exports.Env = exports.EmbeddedAssets = exports.ConnectionStatus = exports.Binaries = exports.Android = exports.IosCodeSigning = exports.Detach = void 0;

function _sourceMapSupport() {
  const data = require("source-map-support");

  _sourceMapSupport = function () {
    return data;
  };

  return data;
}

function _Analytics() {
  const data = _interopRequireDefault(require("./Analytics"));

  _Analytics = function () {
    return data;
  };

  return data;
}

function Android() {
  const data = _interopRequireWildcard(require("./Android"));

  Android = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Android", {
  enumerable: true,
  get: function () {
    return Android();
  }
});

function _Api() {
  const data = _interopRequireDefault(require("./Api"));

  _Api = function () {
    return data;
  };

  return data;
}

function _ApiV() {
  const data = _interopRequireDefault(require("./ApiV2"));

  _ApiV = function () {
    return data;
  };

  return data;
}

function Binaries() {
  const data = _interopRequireWildcard(require("./Binaries"));

  Binaries = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Binaries", {
  enumerable: true,
  get: function () {
    return Binaries();
  }
});

function _Config() {
  const data = _interopRequireDefault(require("./Config"));

  _Config = function () {
    return data;
  };

  return data;
}

function ConnectionStatus() {
  const data = _interopRequireWildcard(require("./ConnectionStatus"));

  ConnectionStatus = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "ConnectionStatus", {
  enumerable: true,
  get: function () {
    return ConnectionStatus();
  }
});

function EmbeddedAssets() {
  const data = _interopRequireWildcard(require("./EmbeddedAssets"));

  EmbeddedAssets = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "EmbeddedAssets", {
  enumerable: true,
  get: function () {
    return EmbeddedAssets();
  }
});

function Env() {
  const data = _interopRequireWildcard(require("./Env"));

  Env = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Env", {
  enumerable: true,
  get: function () {
    return Env();
  }
});

function _ErrorCode() {
  const data = require("./ErrorCode");

  _ErrorCode = function () {
    return data;
  };

  return data;
}

function Exp() {
  const data = _interopRequireWildcard(require("./Exp"));

  Exp = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Exp", {
  enumerable: true,
  get: function () {
    return Exp();
  }
});

function _Logger() {
  const data = _interopRequireDefault(require("./Logger"));

  _Logger = function () {
    return data;
  };

  return data;
}

function _NotificationCode() {
  const data = _interopRequireDefault(require("./NotificationCode"));

  _NotificationCode = function () {
    return data;
  };

  return data;
}

function Project() {
  const data = _interopRequireWildcard(require("./Project"));

  Project = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Project", {
  enumerable: true,
  get: function () {
    return Project();
  }
});

function ProjectAssets() {
  const data = _interopRequireWildcard(require("./ProjectAssets"));

  ProjectAssets = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "ProjectAssets", {
  enumerable: true,
  get: function () {
    return ProjectAssets();
  }
});

function ProjectSettings() {
  const data = _interopRequireWildcard(require("./ProjectSettings"));

  ProjectSettings = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "ProjectSettings", {
  enumerable: true,
  get: function () {
    return ProjectSettings();
  }
});

function Prompts() {
  const data = _interopRequireWildcard(require("./Prompts"));

  Prompts = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Prompts", {
  enumerable: true,
  get: function () {
    return Prompts();
  }
});

function SimControl() {
  const data = _interopRequireWildcard(require("./SimControl"));

  SimControl = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "SimControl", {
  enumerable: true,
  get: function () {
    return SimControl();
  }
});

function Simulator() {
  const data = _interopRequireWildcard(require("./Simulator"));

  Simulator = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Simulator", {
  enumerable: true,
  get: function () {
    return Simulator();
  }
});

function StandaloneBuild() {
  const data = _interopRequireWildcard(require("./StandaloneBuild"));

  StandaloneBuild = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "StandaloneBuild", {
  enumerable: true,
  get: function () {
    return StandaloneBuild();
  }
});

function ThirdParty() {
  const data = _interopRequireWildcard(require("./ThirdParty"));

  ThirdParty = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "ThirdParty", {
  enumerable: true,
  get: function () {
    return ThirdParty();
  }
});

function UrlUtils() {
  const data = _interopRequireWildcard(require("./UrlUtils"));

  UrlUtils = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "UrlUtils", {
  enumerable: true,
  get: function () {
    return UrlUtils();
  }
});

function _User() {
  const data = _interopRequireWildcard(require("./User"));

  _User = function () {
    return data;
  };

  return data;
}

function _UserSettings() {
  const data = _interopRequireDefault(require("./UserSettings"));

  _UserSettings = function () {
    return data;
  };

  return data;
}

function Versions() {
  const data = _interopRequireWildcard(require("./Versions"));

  Versions = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Versions", {
  enumerable: true,
  get: function () {
    return Versions();
  }
});

function Webpack() {
  const data = _interopRequireWildcard(require("./Webpack"));

  Webpack = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Webpack", {
  enumerable: true,
  get: function () {
    return Webpack();
  }
});

function _XDLError() {
  const data = _interopRequireDefault(require("./XDLError"));

  _XDLError = function () {
    return data;
  };

  return data;
}

function AndroidCredentials() {
  const data = _interopRequireWildcard(require("./credentials/AndroidCredentials"));

  AndroidCredentials = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "AndroidCredentials", {
  enumerable: true,
  get: function () {
    return AndroidCredentials();
  }
});

function PKCS12Utils() {
  const data = _interopRequireWildcard(require("./detach/PKCS12Utils"));

  PKCS12Utils = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "PKCS12Utils", {
  enumerable: true,
  get: function () {
    return PKCS12Utils();
  }
});

function _PackagerLogsStream() {
  const data = _interopRequireWildcard(require("./logs/PackagerLogsStream"));

  _PackagerLogsStream = function () {
    return data;
  };

  return data;
}

function Doctor() {
  const data = _interopRequireWildcard(require("./project/Doctor"));

  Doctor = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "Doctor", {
  enumerable: true,
  get: function () {
    return Doctor();
  }
});

function ProjectUtils() {
  const data = _interopRequireWildcard(require("./project/ProjectUtils"));

  ProjectUtils = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "ProjectUtils", {
  enumerable: true,
  get: function () {
    return ProjectUtils();
  }
});

function ModuleVersion() {
  const data = _interopRequireWildcard(require("./tools/ModuleVersion"));

  ModuleVersion = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "ModuleVersion", {
  enumerable: true,
  get: function () {
    return ModuleVersion();
  }
});

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const Detach = require('./detach/Detach.js');

exports.Detach = Detach;

const IosCodeSigning = require('./detach/IosCodeSigning.js');

exports.IosCodeSigning = IosCodeSigning;

if (process.env.NODE_ENV !== 'production' && process.env.NODE_ENV !== 'test') {
  (0, _sourceMapSupport().install)();
}
